### PR TITLE
Cleanup rule printing

### PR DESCRIPTION
**Pull request by Julow**
Refactor printing code to use `Format` features, making it more robust and easier to change.

Rebased on top of https://github.com/realworldocaml/mdx/pull/162

### CODE REVIEW

**Comment by Julow** [comment]
Rebased on master and ready for reviews.

File: bin/rule.ml [hunk 1]
```diff
@@ -142,7 +144,8 @@ let run () md_file section direction prelude prelude_str root =
     let options =
       List.map (Fmt.to_to_string pp_prelude) prelude @
       List.map (Fmt.to_to_string pp_prelude_str) prelude_str @
-      [Fmt.to_to_string pp_direction direction]
+      [Fmt.to_to_string pp_direction direction] @
+      (match section with Some s -> [Fmt.(to_to_string pp_section) s] | None -> [])
```
Comment (samoht) [inline bin/rule.ml]: why is this needed?

**Comment by samoht** [comment]
It's a good idea to simplify that part of the code -- I just don't understand why the patch adding the section is part of the changesets. I guess this is fixing a different issue?

File: bin/rule.ml (hunk 1 above)
Comment (Julow) [reply bin/rule.ml]: That's unrelated.
The `-s` option is used by `rule` but not passed to `test`, that will not work if ignored sections need extra deps.
Other options are correctly passed (`prelude`, `direction`, `root`...).

Should I move this commit to an other PR?

Comment (samoht) [reply bin/rule.ml]: yes please, let's keep that PR focused on cleaning up the code without any bug fixes.

Comment (Julow) [reply bin/rule.ml]: I moved that commit to https://github.com/realworldocaml/mdx/pull/176

**Comment by CraigFe** [comment]
I feel that we could do with direct tests of the `ocaml-mdx rule` command (i.e. generated rules which are _not_  for propping up other tests). This would be enabled by #178.

**Comment by Julow** [comment]
I agree testing that way is not great. We still benefit from generating test rules but we need an other way to test every features of `ocaml-mdx rule`.

File: bin/rule.ml [hunk 2]
```diff
@@ -29,9 +29,18 @@ let prepend_root r b = match r with
   | None   -> b
   | Some r -> Filename.concat r b
 
+let pp_dep fmt = function
+  | `Named (name, path) -> Fmt.pf fmt "(:%s %s)" name path
```
Comment (NathanReb) [inline bin/rule.ml]: Is there any particular reason to use polymorphic variants here? I think I'd prefer we define types for dependencies and actions here.

Comment (Julow) [reply bin/rule.ml]: That's because these values are created a few lines below and directly passed to this function. So it's not worth the overhead of defining the type.
What do you think ?

Comment (NathanReb) [reply bin/rule.ml]: I think that if we extract proper types, move those to the library we can then unit test how they are printed rather than using end to end testing for that.

We can probably do that separately though. Let's keep it like this for now, it's already an improvement!
